### ISSUE TITLE
network interception with `Fetch.enable` breaks cloudflare

### ISSUE THREAD

**Issue by milahu** [kaliiiiiiiiii/Selenium-Driverless#123]
im trying to capture all responses as described in [readme#use-events](https://github.com/kaliiiiiiiiii/Selenium-Driverless#use-events)

cloudflare says

> Please unblock challenges.cloudflare.com to proceed.

chrome shows a warning in the address bar

> your connection to this site is not secure

fixed by adding `options.add_argument("--disable-web-security")`
to [don't enforce the same-origin policy](https://peter.sh/experiments/chromium-command-line-switches/#disable-web-security)

<details>
<summary>
test_selenium_driverless.py
</summary>

```py
#!/usr/bin/env python3

import asyncio
import base64
import sys
import time
import traceback

from cdp_socket.exceptions import CDPError

from selenium_driverless import webdriver


async def on_request(params, global_conn):

    url = params["request"]["url"]
    _params = {"requestId": params['requestId']}
    if params.get('responseStatusCode') in [301, 302, 303, 307, 308]:
        # redirected request
        return await global_conn.execute_cdp_cmd("Fetch.continueResponse", _params)
    else:
        try:
            body = await global_conn.execute_cdp_cmd("Fetch.getResponseBody", _params, timeout=1)
        except CDPError as e:
            if e.code == -32000 and e.message == 'Can only get response body on requests captured after headers received.':
                print(params, "\n", file=sys.stderr)
                traceback.print_exc()
                await global_conn.execute_cdp_cmd("Fetch.continueResponse", _params)
            else:
                raise e
        else:
            start = time.monotonic()
            body_decoded = base64.b64decode(body['body'])

            # modify body here

            body_modified = base64.b64encode(body_decoded).decode("ascii")
            fulfill_params = {"responseCode": 200, "body": body_modified}
            fulfill_params.update(_params)
            _time = time.monotonic() - start
            if _time > 0.01:
                print(f"decoding took long: {_time} s")
            await global_conn.execute_cdp_cmd("Fetch.fulfillRequest", fulfill_params)
            print("Mocked response", url)


async def main():
    options = webdriver.ChromeOptions()
    options.add_argument("--window-size=500,900")
    # fix: please unblock challenges.cloudflare.com to proceed
    # Don't enforce the same-origin policy
    options.add_argument("--disable-web-security")
    async with webdriver.Chrome(options=options, max_ws_size=2 ** 30) as driver:
        driver.base_target.socket.on_closed.append(lambda code, reason: print(f"chrome exited"))
        global_conn = driver.base_target
        await driver.get("about:blank")
        await global_conn.execute_cdp_cmd("Fetch.enable", cmd_args={"patterns": [{"requestStage": "Response", "urlPattern":"*"}]})
        await global_conn.add_cdp_listener("Fetch.requestPaused", lambda data: on_request(data, global_conn))
        await driver.get(
            #'https://wikipedia.org',
            "https://nowsecure.nl/#relax", # test cloudflare
            timeout=60, wait_load=False)
        while True:
            #time.sleep(10) # no. cloudflare would hang
            await asyncio.sleep(10)


asyncio.run(main())
```

</details>

**Comment by kaliiiiiiiiii** [comment]
I can confirm this. However, I suspect this to be a timing leak and cloudfare therefore sending a 403 back=> not really a way to fix.
![image](https://github.com/kaliiiiiiiiii/Selenium-Driverless/assets/89038706/1d1f11f4-e068-4136-9b83-f30ea6e2caab)
![image](https://github.com/kaliiiiiiiiii/Selenium-Driverless/assets/89038706/ef72c7ab-7167-4cac-9b44-eba409d1b3ae)

@milahu or any other thoughts//ideas on that?

**Comment by kaliiiiiiiiii** [comment]
> The problem is that one of Cloudflare's engineers is watching this repository... :)

@juhacz
Likely, yes.

#### Soo in case some @cloudfare staff is reading this:
Why not hire me directly instead of needing someone to analyse & understand the code on here ? :)

**Comment by juhacz** [comment]
@kaliiiiiiiiii Because we need people like you more :) I Suggest creating a profile at https://www.buymeacoffee.com/ I think people will confirm my words :)

**Comment by milahu** [comment]
> I suspect this to be a timing leak

you mean the python response handler is too slow?

or maybe the continueResponse/fulfillRequest logic has a bug

but yeah, it seems to be a new problem
with the error message
"Please unblock challenges.cloudflare.com to proceed."
i only find [a tapatalk.com thread](https://www.tapatalk.com/groups/germanrailfr/please-unblock-challenges-cloudflare-com-to-procee-t11165.html) from 2023-10-30 with no solution

> any other thoughts//ideas on that?

so far i used the "export HAR" function of chrome devtools network
but that is slower than capturing the live traffic

the exported HAR file does not include the bodies of binary responses
which is actually good for large binaries
i dont want to store a 1GB response in RAM
but let chrome write them to the filesystem

chromium is open source, so it should be easy to find
how the "record network log" command works

an alternative would be a local http proxy
i guess `Fetch.enable` also works with a http proxy inside of chrome
and maybe that proxy is visible to cloudflare

in the long term, they will replace captchas with government ID logins
and to bypass that, we will need p2p scraping tools...

**Comment by kaliiiiiiiiii** [comment]
> @kaliiiiiiiiii Because we need people like you more :) I Suggest creating a profile at https://www.buymeacoffee.com/ I think people will confirm my words :)

@juhacz
added:) https://github.com/kaliiiiiiiiii#support-me

**Comment by milahu** [comment]
> chromium is open source, so it should be easy to find
> how the "record network log" command works




[chromium/src/third_party/devtools-frontend/src/front_end/panels/network/network-meta.ts](https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/panels/network/network-meta.ts;l=199)

UIStrings.recordNetworkLog

```ts
UI.ActionRegistration.registerActionExtension({
  actionId: 'network.toggle-recording',
  category: UI.ActionRegistration.ActionCategory.NETWORK,
  iconClass: UI.ActionRegistration.IconClass.START_RECORDING,
  toggleable: true,
  toggledIconClass: UI.ActionRegistration.IconClass.STOP_RECORDING,
  toggleWithRedColor: true,
  contextTypes() {
    return maybeRetrieveContextTypes(Network => [Network.NetworkPanel.NetworkPanel]);
  },
  async loadActionDelegate() {
    const Network = await loadNetworkModule();
    return new Network.NetworkPanel.ActionDelegate();
  },
  options: [
    {
      value: true,
      title: i18nLazyString(UIStrings.recordNetworkLog),
    },
    {
      value: false,
      title: i18nLazyString(UIStrings.stopRecordingNetworkLog),
    },
  ],
```

[chromium/src/third_party/devtools-frontend/src/front_end/panels/network/NetworkPanel.ts](https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/panels/network/NetworkPanel.ts;l=933)

network.toggle-recording

```ts
export class ActionDelegate implements UI.ActionRegistration.ActionDelegate {
  handleAction(context: UI.Context.Context, actionId: string): boolean {
    const panel = context.flavor(NetworkPanel);
    if (panel === null) {
      return false;
    }
    switch (actionId) {
      case 'network.toggle-recording': {
        panel.toggleRecord(!panel.recordLogSetting.get());
        return true;
      }
```

panel.toggleRecord

```ts
  toggleRecord(toggled: boolean): void {
    this.toggleRecordAction.setToggled(toggled);
    if (this.recordLogSetting.get() !== toggled) {
      this.recordLogSetting.set(toggled);
    }

    this.networkLogView.setRecording(toggled);
    if (!toggled && this.filmStripRecorder) {
      this.filmStripRecorder.stopRecording(this.filmStripAvailable.bind(this));
    }
  }
```

toggleRecordAction.setToggled

```ts
    this.toggleRecordAction = UI.ActionRegistry.ActionRegistry.instance().getAction('network.toggle-recording');
```

[chromium/src/third_party/devtools-frontend/src/front_end/panels/network/NetworkLogView.ts](https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/panels/network/NetworkLogView.ts;l=926)

networkLogView.setRecording

```ts
  setRecording(recording: boolean): void {
    this.recording = recording;
    this.updateSummaryBar();
  }
```

hmm, i dont see it yet
i guess "record network log" will set some internal setting in the chrome process
to persist the cache of network requests, and to give read-access to devtools

**Comment by kaliiiiiiiiii** [comment]
@milahu 

> you mean the python response handler is too slow?

yep or maybe even the interception at C++ Chromium is to slow over a single websocket.
1. Long-term workaround here would be ausing smth like [selenium-wire](https://github.com/wkeeling/selenium-wire), this however requires some development, to fix th SSL pinning.

> or maybe the [continueResponse](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-continueResponse)/[fulfillRequest](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-fulfillRequest) logic has a bug (note: continueResponse is experimental)

Yep there for sure are some bugs. What I as well could think of is that maybe some iframes don't get intercepted correctly, and therefore have a detectable difference to the main frame.

> so far i used the "export HAR" function of chrome devtools network but that is slower than capturing the live traffic

Yep that works as well of course, however more a workaround:)
> an alternative would be a local http proxy i guess `Fetch.enable` also works with a http proxy inside of chrome and maybe that proxy is visible to cloudflare

See `1.`
I assumed chrome intercepts directly between `frames | boringssl` and doesn't tunnel it through a proxy after boringssl.
Maybe we can find some source-code on that?

another thing to try is

2. `Network.setRequestInterception` (deprecaded tho). 

Soo feel free to share a POC & status if you try that

**Comment by milahu** [comment]
> interception

for my use case, i dont need any active interception of requests/responses
i just need a passive live-stream of http traffic

**Comment by kaliiiiiiiiii** [comment]
> > an alternative would be a local http proxy
> 
> [selenium-wire](https://github.com/wkeeling/selenium-wire) uses a patched version of [mitmproxy](https://github.com/mitmproxy/mitmproxy) as http proxy
> 
> this also allows for active network interception without `chromium --disable-web-security` because we can tell chromium to trust the proxy's [certificate](https://github.com/wkeeling/selenium-wire#certificates)

still pretty sure the SSL/TLS fingerprint doesn't match to chrome as it doesn't use boringssl tho. see https://github.com/wkeeling/selenium-wire/issues/215#issuecomment-794362654

**Comment by kaliiiiiiiiii** [comment]
Interesting note here that:
```python
from cdp_socket.utils.utils import launch_chrome, random_port
from cdp_socket.socket import CDPSocket
import os
import asyncio

global sock1


async def on_resumed(params):
    global sock1
    await sock1.exec("Fetch.continueRequest", {"requestId": params['requestId']})
    print(params["request"]["url"])


async def main():
    global sock1
    PORT = random_port()
    process = launch_chrome(PORT)

    async with CDPSocket(PORT) as base_socket:
        targets = await base_socket.targets
        target = targets[0]
        sock1 = await base_socket.get_socket(target)
        await sock1.exec("Network.clearBrowserCookies")
        await sock1.exec("Fetch.enable")
        sock1.add_listener("Fetch.requestPaused", on_resumed)
        await sock1.exec("Page.navigate", {"url": "https://nowsecure.nl#relax"})
        await asyncio.sleep(5)

    os.kill(process.pid, 15)


asyncio.run(main())
```
works just fine

**Comment by milahu** [comment]
> works just fine

this works for requests, but not for responses
because `Fetch.getResponseBody` always throws `CDPError -32000`

<details>
<summary>
test.py
</summary>

```py
#!/usr/bin/env python3

# https://github.com/kaliiiiiiiiii/Selenium-Driverless/issues/123#issuecomment-1858803756

from cdp_socket.utils.utils import launch_chrome, random_port
from cdp_socket.socket import CDPSocket
from cdp_socket.exceptions import CDPError

import os
import asyncio
import json
import base64
import sys
import time
import traceback

global sock1


async def on_request_paused(params):
    global sock1

    url = params["request"]["url"]
    url_clean = url.split("?")[0]
    if len(url_clean) > 60:
        url_clean = url_clean[:60] + "..."
    _params = {"requestId": params['requestId']}
    #if params.get('responseStatusCode') in [301, 302, 303, 307, 308]:
    #    # redirected request
    #    return await sock1.exec("Fetch.continueResponse", _params)
    try:
        #print("Fetch.getResponseBody ...", url_clean)
        body = await sock1.exec("Fetch.getResponseBody", _params, timeout=30)
    except CDPError as e:
        #print("Fetch.getResponseBody CDPError", url_clean)
        if e.code == -32000:
            # Can only get response body on HeadersReceived pattern matched requests.
            print("Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse", url_clean)
            #print("Fetch.continueResponse ...", url_clean)
            res = await sock1.exec("Fetch.continueResponse", _params, timeout=30)
            #print("Fetch.continueResponse done", url_clean)
            return res
        else:
            print("Fetch.getResponseBody CDPError raise", url_clean)
            raise e
    else:
        print("Fetch.getResponseBody done", url_clean)
        start = time.monotonic()
        body_decoded = base64.b64decode(body['body'])
        # modify body here
        body_modified = base64.b64encode(body_decoded).decode("ascii")
        fulfill_params = {"responseCode": 200, "body": body_modified}
        fulfill_params.update(_params)
        _time = time.monotonic() - start
        if _time > 0.01:
            print(f"decoding took long: {_time} s")
        print("Fetch.fulfillRequest ...")
        res = await sock1.exec("Fetch.fulfillRequest", fulfill_params, timeout=30)
        print("Fetch.fulfillRequest done", url_clean)
        print("Mocked response", url_clean)
        return res


async def main():
    global sock1
    PORT = random_port()
    process = launch_chrome(PORT)

    async with CDPSocket(PORT) as base_socket:
        targets = await base_socket.targets
        target = targets[0]
        sock1 = await base_socket.get_socket(target)
        await sock1.exec("Network.clearBrowserCookies")
        await sock1.exec("Fetch.enable")
        sock1.add_listener("Fetch.requestPaused", on_request_paused)
        # timeout: fix: asyncio.exceptions.TimeoutError
        await sock1.exec("Page.navigate", {"url": "https://nowsecure.nl#relax"}, timeout=30)
        print("waiting after Page.navigate")
        await asyncio.sleep(5)

    os.kill(process.pid, 30)


asyncio.run(main())
```

</details>

```
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://nowsecure.nl/
waiting after Page.navigate
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://nowsecure.nl/cdn-cgi/styles/challenges.css
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://nowsecure.nl/cdn-cgi/challenge-platform/h/g/orchestr...
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://challenges.cloudflare.com/turnstile/v0/g/74bd6362/ap...
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://nowsecure.nl/favicon.ico
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://nowsecure.nl/cdn-cgi/challenge-platform/h/g/flow/ov1...
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://nowsecure.nl/favicon.ico
Fetch.getResponseBody CDPError -32000 -> Fetch.continueResponse https://challenges.cloudflare.com/cdn-cgi/challenge-platform...
```

similar...
https://github.com/cloud-browser/scrapy-cloud-browser/blob/main/scrapy_cloud_browser/scenarist/page.py

**Comment by milahu** [comment]
`chrome://net-export/` could be useful for passive capturing of traffic

> Click the button to start logging future network activity to a file on disk. The log includes details of network activity from all of Chrome, including incognito and non-incognito tabs, visited URLs, and information about the network configuration

via `chrome://net-internals/`

**Comment by kaliiiiiiiiii** [comment]
Looks like Network.setRequestInterception has the same issues. WOnder tho why it's flaged as "Insecure", eventho the request is over HTTPS
![image](https://github.com/kaliiiiiiiiii/Selenium-Driverless/assets/89038706/6a47a89f-09e7-40dc-ba1d-0c59eba47490)
![image](https://github.com/kaliiiiiiiiii/Selenium-Driverless/assets/89038706/cb5c0bc1-afb0-4ea6-8ae6-807fe70617d3)


<details>

```python
import asyncio
import base64
import sys
import time
import traceback

from cdp_socket.exceptions import CDPError

from selenium_driverless import webdriver


async def on_request(params, global_conn):
    url = params["request"]["url"]
    _params = {"interceptionId": params['interceptionId']}
    if params.get('responseStatusCode') in [301, 302, 303, 307, 308]:
        # redirected request
        return await global_conn.execute_cdp_cmd("Network.continueInterceptedRequest", _params)
    else:
        try:
            body = await global_conn.execute_cdp_cmd("Network.getResponseBodyForInterception", _params, timeout=1)
        except CDPError as e:
            if e.code == -32000 and e.message == 'Can only get response body on requests captured after headers received.':
                print(params, "\n", file=sys.stderr)
                traceback.print_exc()
                await global_conn.execute_cdp_cmd("Fetch.continueResponse", _params)
            else:
                raise e
        else:
            start = time.monotonic()
            body_encoded = base64.b64decode(body['body'])

            # modify body here

            body_modified = base64.b64encode(body_encoded).decode()
            fulfill_params = {"rawResponse": body_modified}
            fulfill_params.update(_params)
            _time = time.monotonic() - start
            if _time > 0.01:
                print(f"decoding took long: {_time} s")
            await global_conn.execute_cdp_cmd("Network.continueInterceptedRequest", fulfill_params)
            print("Mocked response", url)


async def main():
    options = webdriver.ChromeOptions()
    async with webdriver.Chrome(options=options, max_ws_size=2 ** 30) as driver:
        driver.base_target.socket.on_closed.append(lambda code, reason: print(f"chrome exited"))
        global_conn = driver.current_target
        await driver.get("about:blank")
        await global_conn.execute_cdp_cmd("Network.enable", {"maxTotalBufferSize": 1_000_000,  # 1GB
                                                             "maxResourceBufferSize":1_000_000,
                                                             "maxPostDataSize":1_000_000
                                                             })
        await global_conn.execute_cdp_cmd("Network.setRequestInterception", {"patterns":[{"urlPattern":"*", "interceptionStage":"HeadersReceived"}]})
        await global_conn.add_cdp_listener("Network.requestIntercepted", lambda data: on_request(data, global_conn))
        await driver.get(
            'https://nowsecure.nl',
            timeout=60, wait_load=False)
        while True:
            await asyncio.sleep(10)


asyncio.run(main())
```

<details/>

**Comment by milahu** [comment]
> wonder tho why it's flaged as "Insecure", eventho the request is over HTTPS

i guess it uses a local https proxy with a self-signed certificate
without adding that certificate as "trusted cert" to `~/.pki/nssdb/`

**Comment by kaliiiiiiiiii** [comment]
Also interesting here, that local overrides with the chrome devtools just work fine:
![image](https://github.com/kaliiiiiiiiii/Selenium-Driverless/assets/89038706/6ce6c17a-46e5-4161-bf18-eb5faf251852)

> i guess it uses a local https proxy with a self-signed certificate
> without adding that certificate as "trusted cert" to ~/.pki/nssdb/

ahh yep, that makes sense

> but still, this fails to bypass cloudflare

maybe there's a way to detect self-signed certificate usage? If no, it's probably timing or SSL//TLS fingerprinting I guess

I see 2 possible aproaches here:
1. check if we can access that over a chrome extensions (check if existing ones work) @milahu feel free to lmk if you find a workimg one. Getting the source-code & analysing shouldn't be that hard.
2. What if we, instead of mofifying the body binary, point the url to a local webserver?

**Comment by milahu** [comment]
for now i gave up on intercepting requests...
chrome seems to make it really hard, also to provide security against MITM attacks

probably i would try the `frida` route
as i described in https://github.com/wkeeling/selenium-wire/issues/656#issuecomment-1848393185

**Comment by kaliiiiiiiiii** [comment]
> for now i gave up on intercepting requests... chrome seems to make it really hard, also to provide security against MITM attacks
> 
> probably i would try the `frida` route as i described in [wkeeling/selenium-wire#656 (comment)](https://github.com/wkeeling/selenium-wire/issues/656#issuecomment-1848393185)

Well yeah, eventho I assume that the memory manipulation//ddl hooking solutions are specific to:
- chrome versions
- OS
and therefore hard to maintain long-term:/

**Comment by kaliiiiiiiiii** [comment]
At

> `chrome://net-export/` could be useful for passive capturing of traffic
> 
> > Click the button to start logging future network activity to a file on disk. The log includes details of network activity from all of Chrome, including incognito and non-incognito tabs, visited URLs, and information about the network configuration
> 
> via `chrome://net-internals/`

Uhh I think passive capturing works as well with `Fetch.enable` or `Network.setRequestInterception` as long you don't modify the body btw

**Comment by kaliiiiiiiiii** [comment]
Even changing request headers works just fine

![image](https://github.com/kaliiiiiiiiii/Selenium-Driverless/assets/89038706/fbfdd9f4-c1a8-4319-88b1-975f444a3e52)

```python
import asyncio
import base64
import sys
import time
import traceback

from cdp_socket.exceptions import CDPError

from selenium_driverless import webdriver


async def on_request(params, global_conn):
    url = params["request"]["url"]
    _params = {"interceptionId": params['interceptionId']}
    if params.get('responseStatusCode') in [301, 302, 303, 307, 308]:
        # redirected request
        return await global_conn.execute_cdp_cmd("Network.continueInterceptedRequest", _params)
    else:

        fulfill_params = {"headers":params["request"]["headers"]}
        fulfill_params["headers"]["test"] = "Hello World!"
        fulfill_params.update(_params)
        await global_conn.execute_cdp_cmd("Network.continueInterceptedRequest", fulfill_params)
        print(url)


async def main():
    options = webdriver.ChromeOptions()
    async with webdriver.Chrome(options=options, max_ws_size=2 ** 30) as driver:
        driver.base_target.socket.on_closed.append(lambda code, reason: print(f"chrome exited"))
        global_conn = driver.current_target
        await driver.get("about:blank")
        await global_conn.execute_cdp_cmd("Network.enable", {"maxTotalBufferSize": 1_000_000,  # 1GB
                                                             "maxResourceBufferSize": 1_000_000,
                                                             "maxPostDataSize": 1_000_000
                                                             })
        await global_conn.execute_cdp_cmd("Network.setRequestInterception",
                                          {"patterns": [{"urlPattern": "*",
                                                         # "interceptionStage": "HeadersReceived"
                                                         }]})
        await global_conn.add_cdp_listener("Network.requestIntercepted", lambda data: on_request(data, global_conn))
        await driver.get(
            'https://nowsecure.nl',
            timeout=60, wait_load=False)
        while True:
            await asyncio.sleep(10)


asyncio.run(main())
```

**Comment by milahu** [comment]
> `print(url)`

and where is the response body?

**Comment by milahu** [comment]
> and where is the response body?

[Network.getResponseBody](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-getResponseBody)

<details>

```py
#!/usr/bin/env python3

import asyncio
from selenium_driverless import webdriver
from selenium_driverless.types.by import By
import base64

async def main():

    driver = await webdriver.Chrome()
    #await asyncio.sleep(1)

    target = None

    async def requestWillBeSent(args):
        #print("requestWillBeSent", args)
        print("requestWillBeSent", args["request"]["url"])

    async def requestWillBeSentExtraInfo(args):
        print("requestWillBeSentExtraInfo", args)

    async def responseReceived(args):
        # TODO better. get target of this response
        nonlocal target
        #print("responseReceived", args)
        status = args["response"]["status"]
        url = args["response"]["url"]
        _type = args["response"]["headers"]["Content-Type"]

        # TODO better. detect when response data is ready
        # fix: No data found for resource with given identifier
        await asyncio.sleep(1)

        args = {
            "requestId": args["requestId"],
        }
        body = await target.execute_cdp_cmd("Network.getResponseBody", args)
        body = base64.b64decode(body["body"]) if body["base64Encoded"] else body["body"]

        print("responseReceived", status, url, _type, repr(body[:20]) + "...")

    async def responseReceivedExtraInfo(args):
        print("responseReceivedExtraInfo", args)

    async def targetCreated(args):
        print("targetCreated", args)

    async def targetInfoChanged(args):
        #print("targetInfoChanged", args)
        print("targetInfoChanged")

    target = await driver.current_target
    #print("target.id", target.id)

    # enable Target events
    args = {
        "discover": True,
        #"filter": ...
    }
    await target.execute_cdp_cmd("Target.setDiscoverTargets", args)

    await target.add_cdp_listener("Target.targetCreated", targetCreated)
    await target.add_cdp_listener("Target.targetInfoChanged", targetInfoChanged)

    #print("driver.targets", await driver.targets)

    # enable Network events
    args = {
        "maxTotalBufferSize": 1_000_000,  # 1GB
        "maxResourceBufferSize": 1_000_000,
        "maxPostDataSize": 1_000_000
    }
    await target.execute_cdp_cmd("Network.enable", args)

    await target.add_cdp_listener("Network.requestWillBeSent", requestWillBeSent)
    #await target.add_cdp_listener("Network.requestWillBeSentExtraInfo", requestWillBeSentExtraInfo)
    await target.add_cdp_listener("Network.responseReceived", responseReceived)
    #await target.add_cdp_listener("Network.responseReceivedExtraInfo", responseReceivedExtraInfo)



    #await asyncio.sleep(1)

    url = "http://httpbin.org/get"
    print("driver.get", url)
    await driver.get(url)
    await asyncio.sleep(3)

    #print("driver.targets", await driver.targets)

    """
    print("hit enter to close")
    input()
    """

    await driver.close()

asyncio.run(main())
```

example output

```
driver.get http://httpbin.org/get
requestWillBeSent http://httpbin.org/get
targetInfoChanged
requestWillBeSent http://httpbin.org/favicon.ico
responseReceived 200 http://httpbin.org/get application/json '{\n  "args": {}, \n  "'...
responseReceived 404 http://httpbin.org/favicon.ico text/html '<!DOCTYPE HTML PUBLI'...
```

</details>

**Comment by milahu** [comment]
> Please unblock challenges.cloudflare.com to proceed.

this error appears when `Fetch.fulfillRequest` has no response headers

fix:

```py
    async def requestPaused(args):
        # ...
        body = base64.b64encode(body).decode("ascii")
        _args = {
            "requestId": args["requestId"],
            "responseCode": args["responseStatusCode"],
            # fix: Please unblock challenges.cloudflare.com to proceed.
            "responseHeaders": args["responseHeaders"],
            "body": body,
        }
        if args["responseStatusText"] != "":
            # empty string throws "Invalid http status code or phrase"
            _args["responsePhrase"] = args["responseStatusText"]
        await target.execute_cdp_cmd("Fetch.fulfillRequest", _args)
```


> passive capturing works as well with `Fetch.enable` or `Network.setRequestInterception` as long you don't modify the body

im looking for a generic solution, based on streams
so i can handle infinite-size responses without storing the whole response in RAM
and so i can handle streams of events with low latency

- works: [Network.enable](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-enable) and [Fetch.requestPaused](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#event-requestPaused) and [Fetch.takeResponseBodyAsStream](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-takeResponseBodyAsStream) and [Fetch.fulfillRequest](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-fulfillRequest) - this is not perfect, because i cannot send a stream back to chromium, [there is no Fetch.giveResponseBodyAsStream and IO.write](https://groups.google.com/g/chrome-debugging-protocol/c/w65z0cMqgvc). but i can abort the response with [Fetch.failRequest](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-failRequest), which works for file downloads, but not for content that is also needed in the html page. this allows active interception of responses
- broken: [Network.enable](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-enable) and [Network.streamResourceContent](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-streamResourceContent) and [Network.dataReceived](https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-dataReceived) - this is broken in chromium 117, because data is always empty. this would be nice, because it allows passive capturing of infinite streams, and the original response stream is sent to the browser. todo: maybe this works in chromium 119
- deprecated: [Network.setRequestInterception](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-setRequestInterception) and [Network.requestIntercepted](https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-requestIntercepted)
- deprecated: [Network.takeResponseBodyForInterceptionAsStream](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-takeResponseBodyForInterceptionAsStream) is as good as deprecated, because [Network.requestIntercepted](https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-requestIntercepted) is deprecated, which is needed to get `interceptionId`

see also https://github.com/milahu/aiohttp_chromium/tree/main/test/stream-response

**Comment by kaliiiiiiiiii** [comment]
> see also https://github.com/milahu/aiohttp_chromium/tree/main/test/stream-response
>
> feel free to copy/paste/modify these scripts to Selenium-Driverless/examples/

ah yep, thanks. Might be nice if you can keep it up long-term somewhere in your repo for reference


> broken: [Network.enable](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-enable) and [Network.streamResourceContent](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-streamResourceContent) and [Network.dataReceived](https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-dataReceived) - this is broken in chromium 117, because data is always empty.

ah heck, well then `Network` usage should probably be avoided as it's deprecated and more stuff might break in future chrome versions

> > Please unblock challenges.cloudflare.com to proceed.
>
> this error appears when Fetch.fulfillRequest has no response headers
>
> ```python
>     async def requestPaused(args):
>         # ...
>         body = base64.b64encode(body).decode("ascii")
>         _args = {
>             "requestId": args["requestId"],
>             "responseCode": args["responseStatusCode"],
>             # fix: Please unblock challenges.cloudflare.com to proceed.
>             "responseHeaders": args["responseHeaders"],
>             "body": body,
>         }
>         if args["responseStatusText"] != "":
>             # empty string throws "Invalid http status code or phrase"
 >            _args["responsePhrase"] = > args["responseStatusText"]
>         await target.execute_cdp_cmd("Fetch.fulfillRequest", _args)
> ```

Uh nice that we've finally got it working! Great job! 
Wonder, is there any way to optimize `base64.b64encode(body).decode("ascii")` even more btw?

And also, are we sure that `Fetch.enable` intercepts as well:
1. WebWorkers & service-workers
2. cross//OOPIF iframes?
3. background scripts in extensions.

I remember there being `Network.setBypassServiceWorker`, however no idea if it affects `Fetch.enable` as well.

If some still don't get intercepted, maybe target-interception might be considerable, see https://github.com/kaliiiiiiiiii/Selenium-Driverless/blob/4b71a5ab59a193d41eab80ed8f68a66e8ad5c230/tests/target_interception.py . I'm however not sure how reliable it is and how bad the timing leaks are.

**Comment by milahu** [comment]
> then `Network` usage should probably be avoided as it's deprecated and more stuff might break in future chrome versions

[Network.streamResourceContent](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-streamResourceContent) and [Network.dataReceived](https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-dataReceived)
are not deprecated, but experimental
so i expect them to work in newer versions

> is there any way to optimize `base64.b64encode(body).decode("ascii")`

im afraid no... i also would prefer a binary protocol, no base64, no json

base64 is needed for [Fetch.fulfillRequest](https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-fulfillRequest)

> body: string: A response body. If absent, original response body will be used if the request is intercepted at the response stage and empty body will be used if the request is intercepted at the request stage. (Encoded as a base64 string when passed over JSON)

when i pass the body as `bytes` i get

> TypeError: Object of type bytes is not JSON serializable

per [CDP docs](https://chromedevtools.github.io/devtools-protocol/), the only non-JSON endpoint is

> WebSocket `/devtools/page/{targetId}`
> The WebSocket endpoint for the protocol.

> are we sure that `Fetch.enable` intercepts as well

no idea, i dont need these targets

in [Fetch.requestPaused.py](https://github.com/milahu/aiohttp_chromium/blob/main/test/stream-response/Fetch.requestPaused.py) im calling

```py
    target = await driver.current_target
    # ...
    await target.execute_cdp_cmd("Fetch.enable", args)
    await target.add_cdp_listener("Fetch.requestPaused", requestPaused)
```

but this also works with

```py
    await driver.execute_cdp_cmd("Fetch.enable", args)
    await driver.add_cdp_listener("Fetch.requestPaused", requestPaused)
```

then `requestPaused` should be called for all targets

**Comment by kaliiiiiiiiii** [comment]
Also, I'm just thinking about - if we can't stream the responses when intercepting the requests - there's technically a way to detect the timing (if the server responds in chuncks), right?

And even if it would be possible, I suppose there could be a way to setup a server with sepecific chunk timing & size + detect that at JavaScript.

**Comment by milahu** [comment]
aah, now i understand your question

> are we sure that `Fetch.enable` intercepts as well

so ideally, all targets should be intercepted
to add the same latency to all requests

practically, i would avoid this premature optimization
because different latencies can have legitimate reasons
like different cpu loads on different cpu cores

**Comment by kaliiiiiiiiii** [comment]
> ```python
> target = await driver.current_target
> # ...
> await target.execute_cdp_cmd("Fetch.
> ```

yeah ofc - as this will executes cdp on the same target.

I'm not sure if//how `driver.base_target` behaves tbh. I could imagine, that service-worker requests are only covered by base_target. At least for target interception, this is the case.

**Comment by milahu** [comment]
> [Network.streamResourceContent](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-streamResourceContent) and [Network.dataReceived](https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-dataReceived)
> are not deprecated, but experimental
> so i expect them to work in newer versions

bad news: this also fails with chromium 120

maybe this is a bug in `selenium_driverless`?
tomorrow i will port [Network.dataReceived.py](https://github.com/milahu/aiohttp_chromium/blob/main/test/stream-response/Network.dataReceived.py) to `selenium`
i would be surprised if this is a chromium bug

**Comment by milahu** [comment]
> tomorrow i will port [Network.dataReceived.py](https://github.com/milahu/aiohttp_chromium/blob/main/test/stream-response/Network.dataReceived.py) to `selenium`

not possible, because chromedriver does not support the [Network.streamResourceContent](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-streamResourceContent) command

so there is no

```py
await session.execute(devtools.network.stream_resource_content(request_id))
# or
driver.execute("Network.streamResourceContent", {"requestId": request_id})
```

there is only [network.take_response_body_for_interception_as_stream](https://py-cdp.readthedocs.io/en/latest/api/network.html#cdp.network.take_response_body_for_interception_as_stream)

```py
await session.execute(devtools.network.take_response_body_for_interception_as_stream(interception_id))
```

... but that requires an `interception_id`
and there is still no `IO.write` so i cannot send the stream to chromium

see also [Selenium 4: how add event listeners in CDP](https://stackoverflow.com/questions/66227508/selenium-4-0-0-beta-1-how-add-event-listeners-in-cdp)

## CDP is broken by design?

i have the impression that this feature (reading and writing of streams)
is deliberately not implemented by CDP

see also [Fetch.fulfillRequest and (very) long body](https://groups.google.com/g/chrome-debugging-protocol/c/w65z0cMqgvc)

> Unfortunately, there's no streaming support for Fetch network interception at the moment

yeah, totally "unfortunately" and totally "at the moment"

no, i guess this is very deliberate sabotage, to prevent "abusing" chromium as a generic http client
which is pretty much what we are trying to do here...

## dynamic analysis

so... i really tried to avoid this part (because i have zero experience here)
but i will have to use frida to insert hooks into the chromium binary

> for now i gave up on intercepting requests... chrome seems to make it really hard, also to provide security against MITM attacks
> 
> probably i would try the `frida` route as i described in [wkeeling/selenium-wire#656 (comment)](https://github.com/wkeeling/selenium-wire/issues/656#issuecomment-1848393185)

lets see [what tomorrow will bring](https://www.youtube.com/watch?v=N3QVXxquc4g#title=RJD2%20-%20Smoke%20and%20Mirrors) ; )

**Comment by kaliiiiiiiiii** [comment]
> > tomorrow i will port [Network.dataReceived.py](https://github.com/milahu/aiohttp_chromium/blob/main/test/stream-response/Network.dataReceived.py) to `selenium`
> 
> not possible, because chromedriver does not support the [Network.streamResourceContent](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-streamResourceContent) command
> 
> so there is no
> 
> ```python
> await session.execute(devtools.network.stream_resource_content(request_id))
> # or
> driver.execute("Network.streamResourceContent", {"requestId": request_id})
> ```
> 
> there is only [network.take_response_body_for_interception_as_stream](https://py-cdp.readthedocs.io/en/latest/api/network.html#cdp.network.take_response_body_for_interception_as_stream)
> 
> ```python
> await session.execute(devtools.network.take_response_body_for_interception_as_stream(interception_id))
> ```
> 
> ... but that requires an `interception_id` and there is still no `IO.write` so i cannot send the stream to chromium
> 
> see also [Selenium 4: how add event listeners in CDP](https://stackoverflow.com/questions/66227508/selenium-4-0-0-beta-1-how-add-event-listeners-in-cdp)
> 
> ## CDP is broken by design?
> i have the impression that this feature (reading and writing of streams) is deliberately not implemented by CDP

Yeah that might indeed be the case. As well due to security reasons such as streaming all stuff encrypted trough a proxy.
> see also [Fetch.fulfillRequest and (very) long body](https://groups.google.com/g/chrome-debugging-protocol/c/w65z0cMqgvc)
> 
> > Unfortunately, there's no streaming support for Fetch network interception at the moment
> 
> yeah, totally "unfortunately" and totally "at the moment"
> 
> no, i guess this is very deliberate sabotage, to prevent "abusing" chromium as a generic http client which is pretty much what we are trying to do here...

yeah, I guess so
> ## dynamic analysis
> so... i really tried to avoid this part (because i have zero experience here) but i will have to use frida to insert hooks into the chromium binary
> 
> > for now i gave up on intercepting requests... chrome seems to make it really hard, also to provide security against MITM attacks
> > probably i would try the `frida` route as i described in [wkeeling/selenium-wire#656 (comment)](https://github.com/wkeeling/selenium-wire/issues/656#issuecomment-1848393185)
> 
> lets see [what tomorrow will bring](https://www.youtube.com/watch?v=N3QVXxquc4g#title=RJD2%20-%20Smoke%20and%20Mirrors) ; )

well have funn hahe👀 gonna be a pain. Pretty sure Chrome has stuff against that implemented.

**Comment by kaliiiiiiiiii** [comment]
not resolved yet lol

**Comment by milahu** [comment]
well... the original issue is fixed by sending `responseHeaders`

currently i dont have time to implement reading and writing of streams
also i guess this is out-of-scope for `selenium_driverless`
because this is not possible with CDP

**Comment by kaliiiiiiiiii** [comment]
@milahu 
> probably i would try the frida route

Maybe https://github.com/tomer8007/chromium-ipc-sniffer could be a consideration worth👀
screenshot below id 4 years old, some stuff might have changed ofc. 

![](https://raw.githubusercontent.com/tomer8007/chromium-ipc-sniffer/master/screenshots/screenshot_3.png)

**Comment by milahu** [comment]
i would be surprised if that works
the raw HTTP traffic is hidden for better security

> However, this project won't see anything that doesn't go over pipes, which is mostly shared memory IPC:
> 
> - Mojo data pipe contents (raw networking buffers, audio, etc.)

... so the raw HTTP traffic is in shared memory

the most promising method is running chromium in a debugger, either gdb or lldb
but i have to disable sandboxing to set breakpoints on BIO_read and BIO_write
radare is too slow, frida fails to hook the functions
gdb works, but parsing its output is slow, and gdb in python is kinda broken
lldb would be better for interfacing with python (or native code), but its kinda broken...
see also [chromium-capture-http](https://github.com/milahu/aiohttp_chromium/tree/main/test/chromium-capture-http)

but all these are workarounds
and a proper fix would be to implement full http stream support
to fix either [Fetch.requestPaused.py](https://github.com/milahu/aiohttp_chromium/blob/main/test/stream-response/Fetch.requestPaused.py) or [Network.dataReceived.py](https://github.com/milahu/aiohttp_chromium/blob/main/test/stream-response/Network.dataReceived.py)

effectively, this would allow inserting an http proxy
with full control over request and response streams

> its surprising that such a basic feature is missing
> 
> there is Fetch.takeResponseBodyAsStream and IO.read
> but not Fetch.giveResponseBodyAsStream and IO.write
> 
> there is Network.takeResponseBodyForInterceptionAsStream and IO.read
> but not Network.giveResponseBodyForInterceptionAsStream and IO.write

currently this has zero priority for me, i just dont need it

**Comment by ZakariaMQ** [comment]
@kaliiiiiiiiii @milahu
just a little addition from me 
the new headless mode of puppeteer has a nearly regular fingerprint a real Chrome browser
I was able to pass with it so many cloudflare protected website 
and  "**Antoine Vastel, PhD, Head of Research at DataDome**" Admit it in his last interview

adding to puppeteer some custom patches + some mouse movements can be game changer

also, there is a new protocol rather than CDP which is WebDriver BiDi
introduced by google more info here https://developer.chrome.com/blog/webdriver-bidi/

**Comment by kaliiiiiiiiii** [comment]
will be fixed with https://github.com/kaliiiiiiiiii/Selenium-Driverless/blob/dev/src/selenium_driverless/scripts/network_interceptor.py

I'll close this issue when it's released & the documentation is complete

**Comment by kaliiiiiiiiii** [comment]
resolved with https://kaliiiiiiiiii.github.io/Selenium-Driverless/api/RequestInterception/

**Comment by milahu** [comment]
> a proper fix would be to implement full http stream support

nothing new from google
https://issues.chromium.org/issues/332570739

just another feature request
which would be easy to implement, but is ignored as "low priority"
